File: code/Project_4/BT7/tree_traversal.py
from collections import deque

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def build_tree(edges):
    if not edges:
        return None
    
    nodes = {}
    children = set()
    
    for parent, left, right in edges:
        if parent not in nodes:
            nodes[parent] = Node(parent)
        if left != -1:
            if left not in nodes:
                nodes[left] = Node(left)
            nodes[parent].left = nodes[left]
            children.add(left)
        if right != -1:
            if right not in nodes:
                nodes[right] = Node(right)
            nodes[parent].right = nodes[right]
            children.add(right)
    
    # Find root (node that is not a child)
    root_val = None
    for val in nodes:
        if val not in children:
            root_val = val
            break
    
    return nodes[root_val] if root_val else None

def preorder(root):
    if root:
        print(root.val, end=" ")
        preorder(root.left)
        preorder(root.right)

def postorder(root):
    if root:
        postorder(root.left)
        postorder(root.right)
        print(root.val, end=" ")

def levelorder(root):
    if not root:
        return
    q = deque([root])
    while q:
        node = q.popleft()
        print(node.val, end=" ")
        if node.left:
            q.append(node.left)
        if node.right:
            q.append(node.right)

def bottomup(root):
    if not root:
        return
    q = deque([root])
    result = []
    while q:
        node = q.popleft()
        result.append(node.val)
        if node.left:
            q.append(node.left)
        if node.right:
            q.append(node.right)
    result.reverse()
    print(" ".join(map(str, result)))

# Main execution
def main():
    try:
        # Read input from file
        with open('tree_sample.inp', 'r') as f:
            n = int(f.readline().strip())
            edges = []
            for _ in range(n):
                line = list(map(int, f.readline().strip().split()))
                edges.append(line)
        
        # Build tree
        root = build_tree(edges)
        
        # Execute all traversals
        print("Preorder:")
        preorder(root)
        print("\nDone")
        
        print("Postorder:")
        postorder(root)
        print("\nDone")
        
        print("Level-order (Top-down):")
        levelorder(root)
        print("\nDone")
        
        print("Bottom-up:")
        bottomup(root)
        print("\nDone")
        
    except FileNotFoundError:
        print("Error: tree_sample.inp not found")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: code/Project_4/BT7/tree_traversal.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <queue>
#include <stack>
#include <vector>
using namespace std;

struct Node {
    int val;
    Node* left;
    Node* right;
    Node(int x) : val(x), left(nullptr), right(nullptr) {}
};

Node* build_tree(int edges[][3], int n) {
    if (n == 0) return nullptr;
    
    map<int, Node*> nodes;
    set<int> children;
    
    for (int i = 0; i < n; i++) {
        int parent = edges[i][0];
        int left = edges[i][1];
        int right = edges[i][2];
        
        if (nodes.find(parent) == nodes.end())
            nodes[parent] = new Node(parent);
            
        if (left != -1) {
            if (nodes.find(left) == nodes.end())
                nodes[left] = new Node(left);
            nodes[parent]->left = nodes[left];
            children.insert(left);
        }
        
        if (right != -1) {
            if (nodes.find(right) == nodes.end())
                nodes[right] = new Node(right);
            nodes[parent]->right = nodes[right];
            children.insert(right);
        }
    }
    
    // Find root
    for (auto& pair : nodes) {
        if (children.find(pair.first) == children.end()) {
            return pair.second;
        }
    }
    return nullptr;
}

void preorder(Node* root) {
    if (root) {
        cout << root->val << " ";
        preorder(root->left);
        preorder(root->right);
    }
}

void postorder(Node* root) {
    if (root) {
        postorder(root->left);
        postorder(root->right);
        cout << root->val << " ";
    }
}

void levelorder(Node* root) {
    if (!root) return;
    queue<Node*> q;
    q.push(root);
    while (!q.empty()) {
        Node* node = q.front();
        q.pop();
        cout << node->val << " ";
        if (node->left) q.push(node->left);
        if (node->right) q.push(node->right);
    }
}

void bottomup(Node* root) {
    if (!root) return;
    queue<Node*> q;
    stack<int> s;
    q.push(root);
    while (!q.empty()) {
        Node* node = q.front();
        q.pop();
        s.push(node->val);
        if (node->left) q.push(node->left);
        if (node->right) q.push(node->right);
    }
    while (!s.empty()) {
        cout << s.top() << " ";
        s.pop();
    }
}

int main() {
    ifstream infile("tree_sample.inp");
    if (!infile.is_open()) {
        cout << "Error: tree_sample.inp not found\n";
        return 1;
    }
    
    int n;
    infile >> n;
    int edges[1000][3];
    
    for (int i = 0; i < n; i++) {
        infile >> edges[i][0] >> edges[i][1] >> edges[i][2];
    }
    
    Node* root = build_tree(edges, n);
    
    // Execute all traversals
    cout << "Preorder:\n";
    preorder(root);
    cout << "\nDone\n";
    
    cout << "Postorder:\n";
    postorder(root);
    cout << "\nDone\n";
    
    cout << "Level-order (Top-down):\n";
    levelorder(root);
    cout << "\nDone\n";
    
    cout << "Bottom-up:\n";
    bottomup(root);
    cout << "\nDone\n";
    
    return 0;
}